Readme.md
# fib-kv

[![NPM version](https://img.shields.io/npm/v/fib-kv.svg)](https://www.npmjs.org/package/fib-kv)
[![Build Status](https://travis-ci.org/fibjs/fib-kv.svg)](https://travis-ci.org/fibjs/fib-kv)
[![Build status](https://ci.appveyor.com/api/projects/status/they3hi2gq01vrev?svg=true)](https://ci.appveyor.com/project/richardo2016/fib-kv)

## Introduction

general key-value store on sql/level/mongo/redis for fibjs.

## Install

```sh
npm install fib-kv [--save]
```

## Test

```sh
npm run ci
```

## Creating a key-value store

### var kvs = new kv(conn, opts);

```JavaScript
conn: database connection or connection factory like fib-pool
opts: kvs options
```

| opts                 | default | object/Map | LruCache | LevelDB | Redis | MongoDB | SQLite/MySQL |
|----------------------|---------|------------|----------|---------|-------|---------|--------------|
| table_name           |   "kvs" | x          | x        | x       | √     | √       | √            |
| key_name             |     "k" | x          | x        | x       | x     | √       | √            |
| value_name           |     "v" | x          | x        | x       | x     | √       | √            |
| key_size             |      32 | x          | x        | x       | x     | x       | √            |
| value_size           |     256 | x          | x        | x       | x     | x       | √            |
| cleanup_interval(ms) |   60000 | x          | x        | x       | x     | x       | √            |
| timeout(ms)          |       0 | x          | √        | x       | √     | √       | √            |
| prefix               |      "" | √          | √        | √       | √     | √       | √            |
| cache                |   false | √          | √        | √       | √     | √       | √            |
| cache_size           |   65536 | √          | √        | √       | √     | √       | √            |
| cache_timeout(ms)    |   60000 | √          | √        | √       | √     | √       | √            |
| sql_value_type       |   60000 | x          | x        | x       | x     | x       | √            |

**NOTICE** sql_value_type is one of below:
- TINYBLOB
- BLOB
- MEDIUMBLOB
- LONGBLOB
- TINYTEXT
- TEXT
- MEDIUMTEXT
- LONGTEXT
- VARCHAR

A key will not expire if `timeout` is less than or equal to 0.

#### Simple example (memory backend).
```JavaScript
var kv = require("fib-kv");

var kvs = new kv({
    "a": 100,
    "b": 200
});
```

#### LevelDB backend.
```JavaScript
var kv = require("fib-kv");
var db = require("db");

var kvs = new kv(db.openLevelDB("test.ldb"));
```

#### fib-pool backend.
```JavaScript
var kv = require("fib-kv");
var db = require("db");
var pool = require("fib-pool");

var kvs = new kv(pool(() => db.openLevelDB("test.ldb")));
```

## Methods

### kvs.setup()
setup the backend database.

### v = kvs.get(k)
returns the value stored for that key.

### kvs.set(k, v)
stores a key-value pair.

### e = kvs.has(k)
returns whether a key is set on the store.

### e = kvs.keys()
returns all keys in the store.

### e = kvs.renew(k)
renews TTL for an unexpired key.

### kvs.remove(k)
deletes a key-value pair by key.

### e = kvs.cache_has(k)
returns whether a key is set on the cache.

### kvs.cache_clear()
removes all the key-value pairs in the cache.


CHANGELOG.md

v1.3.3 / 2022-11-21
==================

  * feat: better types, support custom_backend

v1.3.2 / 2022-11-20
===================

  * Release v1.3.2
  * chore: update deps, adapt to fib-pool's types.

v1.3.1 / 2019-04-26
===================

  * Release v1.3.1
  * upgrade dependencies.
  * update typo.
  * [README.md] add badgets
  * upgrade project structure.
  * upgrade dependencies version.
  * add ref in .ts
  * Merge pull request #6 from richardo2016/master

v1.3.0 / 2018-07-08
===================

  * Release v1.3.0
  * add test case for sql_value_type.
  * better types.
  * Merge pull request #5 from richardo2016/master
  * all joint-sql to string template.
  * support opts.sql_value_type
  * Merge pull request #4 from richardo2016/master

v1.2.0 / 2018-07-05
===================

  * Release v1.2.0
  * refactor part with typescript, add ci, and add test cases.
  * 修复 fibjs-0.12 以上兼容问题
  * use ensureIndex to create index.
  * Merge pull request #3 from viemacs/devel
  * [fix] LruCache 'get not exists' returns null instead of undefined now
  * [refactor] improve readability of utils
  * 更新版本号
  * Merge pull request #2 from viemacs/f_expire
  * 增加 timeout 选项和相应的 renew, keys 接口
  * 测试用例增加 full 参数
  * 更新依赖版本
  * 更新 repository 地址
  * redis 支持根存储，支持 lrucache
  * 修复 prefix 缺省值
  * 修复 api 文档错误
  * 增加 fib-pool 示例
  * 重构代码，支持 fib-pool
  * 完善 readme
  * 修复 opts 为空时的错误
  * 增加忽略配置，修改版本号
  * 增加 cache 选项
  * init


src/utils.ts
/// <reference path="../@types/index.d.ts" />

const SqlValueTypes: { [key: string /* SQL_VALUE_TYPE */]: true } = {
    'TINYBLOB': true,
    'BLOB': true,
    'MEDIUMBLOB': true,
    'LONGBLOB': true,
    'TINYTEXT' : true,
    'TEXT': true,
    'MEDIUMTEXT': true,
    'LONGTEXT': true,
    'VARCHAR': true
}

export = {
    prefix_key: (key: string, opts: FibKV.FibKVOptions) => opts.prefix !== undefined ? opts.prefix + key : key,
    timeout: (opts: FibKV.FibKVOptions) => opts.timeout !== undefined ? opts.timeout : 0,
    key_name: (opts: FibKV.FibKVOptions) => opts.key_name !== undefined ? opts.key_name : 'k',
    key_size: (opts: FibKV.FibKVOptions) => opts.key_size !== undefined ? opts.key_size : 32,
    pool_name: (opts: FibKV.FibKVOptions) => opts.pool_name !== undefined ? opts.pool_name : '',
    table_name: (opts: FibKV.FibKVOptions) => opts.table_name !== undefined ? opts.table_name : 'kvs',
    value_name: (opts: FibKV.FibKVOptions) => opts.value_name !== undefined ? opts.value_name : 'v',
    value_size: (opts: FibKV.FibKVOptions) => opts.value_size !== undefined ? opts.value_size : 256,
    cache_size: (opts: FibKV.FibKVOptions) => opts.cache_size !== undefined ? opts.cache_size : 65536,
    cache_timeout: (opts: FibKV.FibKVOptions) => opts.cache_timeout !== undefined ? opts.cache_timeout : 60 * 1000,
    cleanup_interval: (opts: FibKV.FibKVOptions) => opts.cleanup_interval !== undefined ? opts.cleanup_interval : 60 * 1000,
    sql_value_type: (opts: FibKV.FibKVOptions) => opts.sql_value_type && SqlValueTypes[opts.sql_value_type] ? opts.sql_value_type : 'VARCHAR'
};
